caper/nb_subproc_thread.py

```diff
@@ -0,0 +1,182 @@
+    def rc(self):
+        """Returns:
+            returncode of subprocess.Popen.poll
+            None if stopped or any other general exception (Exception) occurs
+        """
+        return self._rc
+
+    @property
+    def status(self):
+        """Updated with return value of on_poll() for every polling.
+        Also updated with returnv alue of on_stdout() for every stdout (full line(s)).
+        """
+        return self._status
+
+    @property
+    def ret(self):
+        """Return value updated with return value of on_terminate()
+        None if thread is not done.
+        """
+        return self._ret
+
+    def stop(self):
+        """Subprocess will be teminated after next polling.
+        """
+        self._stop_it = True
+
+    def _popen(
+        self,
+        args,
+        cwd=None,
+        stdin=None,
+        on_poll=None,
+        on_stdout=None,
+        on_terminate=None,
+    ):
+        """Wrapper for subprocess.Popen.
+        stdout/stderr is fixed at subprocess.PIPE/subprocess.STDOUT.
+        i.e. STDERR is always redirected to STDOUT.
+        This thread also has a child thread for non-blocking-streaming of
+        STDOUT.
+
+        Queue implementation for non-blocking STDOUT streaming:
+        https://stackoverflow.com/a/4896288
+        """
+
+        def enqueue_stdout(stdout, q):
+            for line in iter(stdout.readline, b''):
+                q.put(line)
+            stdout.close()
+
+        try:
+            p = Popen(args, stdout=PIPE, stderr=STDOUT, cwd=cwd, stdin=stdin)
+            q = Queue()
+            th_q = Thread(target=enqueue_stdout, args=(p.stdout, q))
+            th_q.daemon = True
+            th_q.start()
+
+            cnt = 0
+            while True:
+                if self._stop_it:
+                    break
+                try:
+                    stdout = q.get_nowait().decode()
+                    if stdout:
+                        self._stdout_list.append(stdout)
+                        if on_stdout:
+                            self._status = on_stdout(stdout)
+                except Empty:
+                    pass
+                except KeyboardInterrupt:
+                    raise
+                if on_poll:
+                    self._status = on_poll(cnt)
+                if p.poll() is not None:
+                    self._rc = p.poll()
+                    break
+                time.sleep(self._poll_interval)
+                cnt += 1
+
+            if self._stop_it:
+                logger.info(
+                    'Stopped subprocess. prev_status={s}'.format(s=self._status)
+                )
+            if self._rc:
+                logger.error(stdout.strip('\n'))
+
+        except CalledProcessError as e:
+            self._rc = e.returncode
+            logger.error(e)
+        except Exception as e:
+            logger.error(e)
+        finally:
+            p.terminate()
+
+        if on_terminate:
+            self._ret = on_terminate()
```

Comment:
Seems hard to test.

Reply:
Removed redundant capturing for `KeyboardInterrupt` in a nested `try`.
It was a bit hard to test since the outer `try` allowed all general exceptions (`Exception`) so removed it and now it only allows `CalledProcessError`.
Also refactored it a bit to add a quiet mode.

All callback functions (on_stdout, on_poll, on_finish), slow/fast polling, `CalledProcessError` and `stop()` are tested. This class is also extensively tested in integration tests (`test_cli_*.py` and `test_cromwell.py`).